docker-compose down

dnephin opened the pull request:
Fixes #2279
Fixes #2277
Fixes #1874

Implements a new `down` command which should be capable of stopping and removing all resources created by `up` (with the correct flags).

I'm not sure how this is going to work with networks, if the network happens to have other containers on it. Does the remove fail?

I think the behaviour of down should be that it logs an error and continues if one of the resources isn't removed, and reports a proper exit code if anything failed.  That isn't implemented yet, but I wanted to get this out before code freeze.

blah238 commented:
Cool, I was just looking for this command, surprised it didn't already exist! :+1:

dnephin commented on compose/project.py:
```diff
@@ -253,6 +253,24 @@ def initialize_volumes(self):
+
+    def remove_volumes(self):
+        for volume in self.volumes:
+            volume.remove()
```
TODO: never remove external volumes

dnephin commented on compose/project.py:
```diff
@@ -253,6 +253,24 @@ def initialize_volumes(self):
+    def down(self, remove_image_type, include_volumes):
+        self.stop()
+        self.remove_stopped(v=include_volumes)
+        self.remove_network()
```
TODO: this needs to be graceful if the remove fails

aanand commented:
Are you planning to finish the TODOs in this PR or separately? In any case, LGTM after rebase.

dnephin commented:
Rebased and made it so that external volumes are never removed. I also fixed the help text to say that only containers and networks are removed by default.

I'm going to leave the "handle network remove failures gracefully" TODO until after networks are fully implemented, I expect behaviour may change slightly.

dnephin commented:
Rebased again, going to merge

albers commented:
@dnephin `docker-compose up` supports services as arguments. Does this make sense for `docker-compose down` as well?

dnephin commented:
I considered that, but I don't think so.  `down` is meant to teardown everything (ex: networks, volumes) which aren't specific to a single service, and will fail if they happen before all the services are stopped and removed.

Instead of trying to handle all those edge cases I opted for not allowing service names. If someone finds a really critical case for it, we can always add it later.

albers commented:
@dnephin Makes sense. I stumbled over this only because of symmetry considerations when working on the completion.
